### ISSUE TITLE
Manage ssh host keys and certs

### ISSUE THREAD

**Issue by arensb** [arensb/ansible-truenas#11]
I'd like to add the ability to manage ssh host keys (e.g., `ssh_host_rsa_key` and `ssh_host_rsa_key.pub`) and especially host key certificates (e.g., `ssh_host_rsa_key-cert.pub`), but I can't find the place in the web console to do so, nor have I found the documentation on how to do so.

The `ssh.config` middleware call returns these keys and certs, but I don't know the best way to update them. By RTFSing, I've found a number of `datastore` calls that can probably do the trick, but since I can't find documentation for this, I suspect this is not the recommended way of doing it.

**Comment by crpb** [comment]
I think you have to updated them in `/data/freenas-v1.db`.`services_ssh`.

If you look at the date in `/usr/local/etc/ssh/ssh_host*` you will likely see the current boot-time. Which service/call writes them out i never bothered to check tho.

**Comment by arensb** [comment]
Yes, that's the file that gets updated by the `datastore` calls that update ssh keys. But I'm looking for the recommended way of updating these keys.
Actually, I'm more interested in updating the certs, but can't find any documentation or guidance on how to do this.

**Comment by darkfiberiru** [comment]
@crpb Can I ask use case of changing host keys?

**Comment by arensb** [comment]
In my case, I'd like to sign the host key with my site CA. In practice, this means generating a key on a secure machine, then installing it on the TrueNAS box. So this could apply to any site where the admins have a key distribution system and want to install their own keys.

**Comment by egandro** [comment]
I would love to have that features!

- change SSH keys (there are multiple ones used for tasks such als truenas/truenas replications)
- SSL certificates for the domains (I use nginx proxy manager to create a *.foo.bar certificate for my internal IPs) and I would love to automate the rollout.

https://www.reddit.com/r/truenas/comments/v0rsfz/ssl_certificates_and_truenas/ << doesn't work for me

**Comment by crpb** [comment]
wouldn't know what has to work on that reddit thing but anyhow to give you a head start

> change SSH keys (there are multiple ones used for tasks such als truenas/truenas replications)

- [manage replication tasks](https://www.truenas.com/docs/api/core_websocket_api.html#replication)
  
- [manage keys](https://www.truenas.com/docs/api/core_websocket_api.html#keychaincredential)


It looks like the credentials in the replication are just a copy
- `midclt call replication.query |jq ' . [] | .ssh_credentials '`
- `midclt call keychaincredential.query | jq ' . [] | select ( .type == "SSH_CREDENTIALS" ) '`

> SSL certificates for the domains (I use nginx proxy manager to create a *.foo.bar certificate for my internal IPs) and I would love to automate the rollout.

- [certificates](https://www.truenas.com/docs/api/scale_websocket_api.html#certificate)

Should all be doable with the [middleware](/arensb/ansible-truenas/plugins/module_utils/middleware.py) as with [sharing_nfs](/arensb/ansible-truenas/plugins/modules/sharing_nfs.py)

**Comment by egandro** [comment]
> Should all be doable with the [middleware](/arensb/ansible-truenas/plugins/module_utils/middleware.py) as with [sharing_nfs](/arensb/ansible-truenas/plugins/modules/sharing_nfs.py)

I started to try to outsmart Truenas by just copying files to `/etc/certificates` - it didn't work (for me) I don't know if there is anything that I have todo to just reread the files. It looks like that there is something in Truenas, that copies the certs from "somewhere else" to this directory during bootup.

**Comment by crpb** [comment]
that won't do :-)
[like above described](https://github.com/arensb/ansible-truenas/issues/11#issuecomment-1927153854) it doesn't help to put files manually anywhere.
this works for some stuff like with the [inetd](https://github.com/crpb/check_mk/blob/main/truenas/check_mk_setup) service where a couple files get touched but you don't wanna do that with the webservice as it's needed for basically anything and you don't wanna break it.
The web-gui(manual) or api(either midclt or rest) are afaik the only sane routes to installing them. But we getting offtopic. This issue is for **SSH** and not the webservice certificates.

**Comment by egandro** [comment]
I think I got a solution for the ssl certificate - https://github.com/mcao2/truenas-cert-updater

**Comment by arensb** [comment]
@egandro Unless I'm missing something, it looks as though that solution only works for the HTTPS web UI (port TCP/443) What I'm looking for is a way to manage the certs used for ssh connections (port TCP/22).

**Comment by egandro** [comment]
> [@egandro](https://github.com/egandro) Unless I'm missing something, it looks as though that solution only works for the HTTPS web UI (port TCP/443) What I'm looking for is a way to manage the certs used for ssh connections (port TCP/22).

I asked for both.

**Comment by crpb** [comment]
@arensb maybe open up an issue in their [jira thingy](https://ixsystems.atlassian.net/jira/software/c/projects/NAS/boards/29) x_X

**Comment by arensb** [comment]
I've done some digging, and found a few things.

`/etc` is mounted as tmpfs, and thus is recreated every time the machine boots. So in order to have keys from `/etc/local/ssh/*` preserved across reboots, they need to be saved to TrueNAS's config database.

Thankfully, there's a service that takes the existing SSH keys and saves them to the config database: `/etc/ix.rc.d/ix_sshd_save_keys`. (It works by using the undocumented `ssh.save_keys` middleware call.)
So the next step should be to enable `ix_sshd_save_keys` so that it runs at boot time. You can see which services are enabled with the `service.query` middleware call (yay!). Unfortunately, the list doesn't include `ix_sshd_save_keys`. And there doesn't appear to be a way to add it (boo!).

(It might be possible to manually add a service entry to the config database, but that seems dangerous.)

Another approach would be to create an init script (under Tasks → Init/Shutdown Scripts). It looks as though preinit scripts run before the `openssh` service, and postinit scripts run after. And I'm pretty sure that `openssh` starts after any ssh keys have been restored from the database. So the thing to do might be to add a postinit script that saves the current set of keys, using `service start ix_sshd_save_keys`.

Currently this Ansible module doesn't support startup scripts, so that should be the next step of development. Looks like the middleware call would be under `initshutdownscript`.

**Comment by crpb** [comment]
Ahh so that private function is callable? i thought that wouldn't work /ö\\

> If you look at https://github.com/truenas/middleware/blob/9c40c3960194ae0acf7a41ce5078794b700455ee/src/middlewared/middlewared/plugins/ssh.py#L104-L122 you can see that `midclt call ssh.update` won't accept those fields as far as i can make out.

**Comment by arensb** [comment]
@crpb Not sure what you're trying to say. You might want to rewrite your comment.

**Comment by arensb** [comment]
I've added a new [release](https://github.com/arensb/ansible-truenas/releases/tag/v1.10.0), v1.10.0, which adds support for init scripts.

The OP problem was to save ssh keys across reboots. To do this, I've used the `initscript` module in the new release to create a job that runs `service ix_sshd_save_keys` at shutdown, to preserve any keys that were changed while the machine was up.

Since this solves the original problem, I'm closing this issue.
